# ==============================================================================
# FIXING COMMIT DIFF
# ==============================================================================
# File: gfx/wr/webrender/src/picture.rs
# Commit: ec05c5d6c331
# Full Hash: ec05c5d6c331b5602f55214a1f08ff0a76f6ad8e
# Author: Glenn Watson <git@intuitionlibrary.com>
# Date: 2022-05-03 09:42:08
# Description:
#   Bug 1765667 - Clamp surface size of content tiles for backdrop-filter r=gfx-reviewers,lsalzman
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D145047
# ==============================================================================

diff -r 40b512ed4dbf -r ec05c5d6c331 gfx/wr/webrender/src/picture.rs
--- a/gfx/wr/webrender/src/picture.rs	Tue May 03 01:38:57 2022 +0000
+++ b/gfx/wr/webrender/src/picture.rs	Tue May 03 01:56:24 2022 +0000
@@ -4827,9 +4827,25 @@
                                         }
                                     }
 
+                                    // We know that we'll never need to sample > 300 device pixels outside the tile
+                                    // for blurring, so clamp the content rect here so that we don't try to allocate
+                                    // a really large surface in the case of a drop-shadow with large offset.
+                                    let max_content_rect = (tile.local_dirty_rect.cast_unit() * device_pixel_scale)
+                                        .inflate(
+                                            MAX_BLUR_RADIUS * BLUR_SAMPLE_SCALE,
+                                            MAX_BLUR_RADIUS * BLUR_SAMPLE_SCALE,
+                                        )
+                                        .round_out()
+                                        .to_i32();
+
                                     let content_device_rect = (local_content_rect.cast_unit() * device_pixel_scale)
                                         .round_out()
                                         .to_i32();
+
+                                    let content_device_rect = content_device_rect
+                                        .intersection(&max_content_rect)
+                                        .expect("bug: no intersection with tile dirty rect");
+
                                     let content_task_size = content_device_rect.size();
                                     let normalized_content_rect = content_task_size.into();
 
